aten/src/ATen/native/cpu/NaiveDilatedConvolution.cpp

```diff
@@ -0,0 +1,67 @@
+#include <ATen/native/NaiveDilatedConvolution.h>
+
+namespace at {
+namespace native {
+namespace {
+
+std::tuple<Tensor, Tensor, Tensor> slow_conv_dilated2d_backward_cpu(
+    const Tensor& grad_output,
+    const Tensor& input,
+    const Tensor& weight,
+    IntArrayRef kernel_size,
+    IntArrayRef stride_size,
+    IntArrayRef pad_size,
+    IntArrayRef dilation_size,
+    const std::array<bool, 3ul> output_mask) {
+  Tensor undefined;
+  internal::slow_conv_dilated_shape_check<2>(
+      input,
+      weight,
+      undefined,
+      grad_output,
+      kernel_size,
+      stride_size,
+      pad_size,
+      dilation_size);
+  auto is_batch = input.dim() == 4;
+  auto options = grad_output.options();
+  // template function assumes batched tensors.  unsqueeze(0) will
+  // insert batch dimension without affecting the original tensor.
+  const Tensor grad_output_ =
+      (is_batch ? grad_output.contiguous()
+                : grad_output.contiguous().unsqueeze(0));
+  const Tensor input_ =
+      (is_batch ? input.contiguous() : input.contiguous().unsqueeze(0));
+  const Tensor weight_ = weight.contiguous();
+  // compute only gradients for which the corresponding output_mask is true:
+  Tensor grad_input =
+      (output_mask[0] ? at::empty(input.sizes(), options) : undefined);
+  Tensor grad_weight =
+      (output_mask[1] ? at::empty(weight.sizes(), options) : undefined);
+  Tensor grad_bias =
+      (output_mask[2] ? at::empty(weight.size(0), options) : undefined);
+  Tensor grad_input_ =
+      (output_mask[0] ? (is_batch ? grad_input : grad_input.unsqueeze(0))
+                      : undefined);
```

Comment:
We should check that all of the Tensors are on CPU.

Previously, slow_conv_dilated2d_backward was a PyTorch dispatcher operation. This meant that if any input tensors are CUDA, then it'll dispatch to the CUDA implementation.

In this PR I think the dispatching logic changed to check if one of the inputs is on CUDA. It's possible that some of the other inputs will be on CUDA, and if we enter this code here with CUDA tensors it'll probably end up segfaulting after it tries to deference a CUDA pointer on CPU.

If that sounds like a correct diagnosis then we should probably also add tests to check that cross-device inputs correctly raise an error message

Reply:
Great catch! Note that there is a check called `slow_conv_dilated_location_check` done at the top of `slow_conv_dilated_all_cpu_template` / `slow_conv_dilated_all_cuda_template` that verifies inputs are all on the same device, so AFAICT it's not possible to trigger this now even with the changes in this PR.

But more generally, I wonder if it's worth having a check at the top of `convolution` / `convolution_backward`.

If it doesn't exist already, I think we should certainly have OpInfo tests that check that a proper error is thrown for the "inputs on different devices" case. I'm not sure but there may be some ops that allow cross-device inputs, and those would have to opt-out of the checks.

Reply:
> Great catch! Note that there is a check called `slow_conv_dilated_location_check` done at the top of `slow_conv_dilated_all_cpu_template` / `slow_conv_dilated_all_cuda_template` that verifies inputs are all on the same device, so AFAICT it's not possible to trigger this now even with the changes in this PR.

Good point, thanks for the catch!

> But more generally, I wonder if it's worth having a check at the top of convolution / convolution_backward that applies to all backends. I don't think any of them support heterogeneous input devices.

It's probably safe to do this until someone comes along with a custom and wants to support heterogeneous input devices lol